src/Features/Core/Portable/EditAndContinue/Remote/RemoteEditAndContinueServiceProxy.cs

```diff
@@ -0,0 +1,449 @@
+            }
+
+            return result.ToImmutable();
+        }
+
+        public async ValueTask<bool> HasChangesAsync(Solution solution, SolutionActiveStatementSpanProvider activeStatementSpanProvider, string sourceFilePath, CancellationToken cancellationToken)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(Workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                return await GetLocalService().HasChangesAsync(solution, activeStatementSpanProvider, sourceFilePath, cancellationToken).ConfigureAwait(false);
+            }
+
+            var result = await client.TryInvokeAsync<IRemoteEditAndContinueService, bool>(
+                solution,
+                (service, solutionInfo, cancellationToken) => service.HasChangesAsync(solutionInfo, sourceFilePath, cancellationToken),
+                callbackTarget: new Callback(activeStatementSpanProvider),
+                cancellationToken).ConfigureAwait(false);
+
+            return result.HasValue ? result.Value : true;
+        }
+
+        public async ValueTask<(SolutionUpdateStatus, ImmutableArray<Deltas>)> EmitSolutionUpdateAsync(Solution solution, SolutionActiveStatementSpanProvider activeStatementSpanProvider, EditAndContinueDiagnosticUpdateSource diagnosticUpdateSource, CancellationToken cancellationToken)
+        {
+            SolutionUpdateStatus status;
+            ImmutableArray<Deltas> deltas;
+            ImmutableArray<DiagnosticData> diagnosticsByProject;
+
+            var client = await RemoteHostClient.TryGetClientAsync(Workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                (status, deltas, diagnosticsByProject) = await GetLocalService().EmitSolutionUpdateAsync(solution, activeStatementSpanProvider, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                var result = await client.TryInvokeAsync<IRemoteEditAndContinueService, (SolutionUpdateStatus, ImmutableArray<Deltas>, ImmutableArray<DiagnosticData>)>(
+                    solution,
+                    (service, solutionInfo, cancellationToken) => service.EmitSolutionUpdateAsync(solutionInfo, cancellationToken),
+                    callbackTarget: new Callback(activeStatementSpanProvider),
+                    cancellationToken).ConfigureAwait(false);
+
+                if (result.HasValue)
+                {
+                    (status, deltas, diagnosticsByProject) = result.Value;
+                }
+                else
+                {
+                    status = SolutionUpdateStatus.Blocked;
+                    deltas = ImmutableArray<Deltas>.Empty;
+                    diagnosticsByProject = ImmutableArray<DiagnosticData>.Empty;
+                }
+            }
+
+            // clear emit/apply diagnostics reported previously:
+            diagnosticUpdateSource.ClearDiagnostics();
```

Comment:
It's a nit, because naming is hard, but I feel like `Proxy` is not ideal given the logic here and in other methods, and perhaps could be renamed in a follow up.

It feels like this class is the real EnC service, and using just the service on its own would be surprising because callers would have to know what extra work to do.

Reply:
I don't think using the real service would be surprising because that service does not have any side-effects. It returns the diagnostics and the caller then is responsible for reporting them. The diagnostic handling here is just a workaround for not having pull diagnostics model and will go away once we remove solution crawler.

That said, I'm open to suggestions for a better name.